src/section/shared_state.rs

```diff
@@ -52,25 +53,34 @@ impl SharedState {
         }
     }
 
-    pub fn update(&mut self, new: Self) {
+    pub fn update(&mut self, new: Self) -> Result<(), RoutingError> {
         if self.handled_genesis_event {
-            log_or_panic!(
-                log::Level::Error,
-                "shared state update - genesis event already handled",
-            );
+            error!("shared state update - genesis event already handled",);
+            return Err(RoutingError::InvalidState);
         }
 
-        if self.our_history.len() > 1 && *self != new {
-            log_or_panic!(
-                log::Level::Error,
-                "shared state update - mismatch: old: {:?} --- new: {:?}",
-                self,
-                new
-            );
+        if self.our_history.len() > 1 {
+            if *self != new {
+                error!(
+                    "shared state update - mismatch: old: {:?} --- new: {:?}",
+                    self, new
+                );
+                return Err(RoutingError::InvalidState);
+            }
+        } else if !new.self_verify() {
+            error!("shared state update - invalid new history: {:?}", new);
+            return Err(RoutingError::InvalidState);
         }
 
         *self = new;
         self.handled_genesis_event = true;
+        Ok(())
+    }
+
+    fn self_verify(&self) -> bool {
+        self.our_history.self_verify()
+            && self.our_members.self_verify()
+            && self.sections.self_verify()
```

Comment:
We are getting there :) But we should also check that all the proofs in all the containers are signed with keys that exists in `self.our_history`. WIth that and https://github.com/maidsafe/routing/pull/2155/files#r452245185, we should be good.

Reply:
that will requires the key history to be pass down across couple of structs.  :(
will have a try